File: python/2-4.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
n,m = map(int, input().split()) 
x,y,direction = map(int, input().split())
o_map = []
g_map = [[0]*n for _ in range(m)] 

dx =[0,1,0,-1]
dy =[-1,0,1,0]

for _ in range(n):
    o_map.append(list(map(int, input().split())))

for i in range(n):
    for j in range(m):
        g_map[i][j] = o_map[i][j]
        
g_map[y][x] = 1
turn_cnt,cnt=0,0

while(True):
    if direction==0:
        direction=3
    else:
        direction+=-1
    turn_cnt+=1

    nx = x+dx[direction]
    ny = y+dy[direction]
    rnx = x-dx[direction]
    rny = y-dy[direction]

    if turn_cnt==4:
        if o_map[rny][rnx]==0 and rnx>=0 and rnx<=n and rny>=0 and rny<=n:
            x,y = rnx,rny
            turn_cnt=0
            cnt+=1
            continue
        else:
            break

    if g_map[ny][nx]==0 and nx>=0 and nx<=n and ny>=0 and ny<=n:
        x,y = nx,ny
        g_map[ny][nx] = 1
        cnt+=1
        turn_cnt=0

print(cnt)

File: c/2-4.c
#include <stdio.h>
#include <string.h>

int main()
{
  int n,m;
  int i,j;
  int x,y,direction;
  int dx[] = {0,1,0,-1};
  int dy[] = {-1,0,1,0};

  scanf("%d %d",&n,&m);
  int map[n][m];
  scanf("%d %d %d",&x,&y,&direction);

  for(i=0;i<n;i++)
    for(j=0;j<m;j++)
      scanf("%d",&map[i][j]);
  int gmap[n][m];
  int count=0;
  int nx, ny, rnx, rny;
  int turn_cnt=0;

  memcpy(gmap,map,sizeof(int)*m*n);
  gmap[y][x]=1;

  while(1)
  {
    if(direction==0)
      direction+=3;
    else
      direction--;
    turn_cnt++;

    nx = x+dx[direction];
    ny = y+dy[direction];
    printf("direction=%d, turn_cnt=%d, (%d %d), count=%d\n",direction, turn_cnt, nx,ny, count);

    if(direction < 2)
    {
      rnx = x+dx[direction+2];
      rny = y+dy[direction+2];

    }
    else
    {
      rnx = x+dx[direction-2];
      rny = y+dy[direction-2];
    }

    if(turn_cnt==4)
    {
      if(map[rny][rnx]==0 && rnx >= 0 && rnx  <= n && rny >= 0 && rny <= n)
      {
        x= rnx;
        y= rny;
        count++;
        turn_cnt=0;
        continue;
      }
      else
      {
        break;
      }
    }

    if(gmap[ny][nx]!=1 && nx >= 0 && nx <=n && ny >=0 && ny<=n)
    {
      gmap[ny][nx]=1;
      turn_cnt=0;
      x=nx;
      y=ny;
      count++;
      printf("direction=%d, turn_cnt=%d, (%d %d), count=%d\n",direction, turn_cnt, nx,ny, count);
      printf("\n");
    }
  }
  printf("%d",count);

  return 0;
}

